Enable/disable logging helpers

amas0 opened the pull request:
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

This is an example implementation to meet #795. It also includes a bug fix pointed out by @skelliest. The primary contribution is the inclusion of the `enable_logging/disable_logging` "functions" at the module level.

Here's an example demonstration of usage:

```python
import cmdstanpy

cmdstanpy.disable_logging()

model = cmdstanpy.CmdStanModel(...)
model.sample(chains=1, show_progress=False)

cmdstanpy.enable_logging()
```

I also tried to be clever and make this same interface work as a context manager, so you can do:

```python
import cmdstanpy

with cmdstanpy.disable_logging():
    model = cmdstanpy.CmdStanModel(...)
    model.sample(chains=1, show_progress=False)
```

I like the idea of this functionality, but the implementation may be stretching the bounds of what is reasonable. For one, the `disable_logging`/`enable_logging` are actually classes so the function call in `cmdstanpy.disable_logging()` actually returns an instance of the `cmdstanpy.utils.logging.disable_logging` class, but this is the only way I could conceive of getting the same interface to function as both a "function" and as a context manager.

The most noticeable result of this is that if a user were to call `cmdstanpy.disable_logging()` in a standalone cell in a Jupyter notebook; they'd see the default `repr` in the cell output. You can get around this by putting a semicolon at the end of the line, which is what I did in the `outputs.rst` docs, but it's more of a band-aid. This bothers me, but I don't see a way around it and also maintain the dual-interface.

Some feedback/thoughts I'd like:

1. Is the juice not worth the squeeze for this dual "function"/context manager?
2. If the dual interface is fine, should the `__repr__`s for these objects be updated so if called in an empty Juypter cell, it's a little cleaner?
3. If not, what would the preferred solution be? Some variations off the top of my head:

* Remove the context manager and implement as a simple function. Probably the most straightforward approach, but will require users to manually re-enable logging after disabling.
* Create separate interfaces for the function and the context manager, something like:
```python
import cmdstanpy

with cmdstanpy.context(logging=False):
  model.sample()
```  
in addition to `cmdstanpy.disable_logging()`

* Remove the _function_ and only have a context manager option. This would encourage only local disabling of logging and the like, while of course leaving open the option to disable logging globally the old fashioned way. 

Probably overthinking what is a relatively minor change, happy to hear some thoughts.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): myself



By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

amas0 committed:
Add enable_logger/disable_logger helpers

amas0 committed:
Fix incorrect check for logger handlers

Credit to skelliest for identifying

amas0 committed:
Update outputs user's guide

amas0 committed:
Add enable/disable_logging to utils.__init__.py all

amas0 committed:
Restructure enable/disable logging

Now have EnableLogging and DisableLogging classes
that are returned by enable_logging and disable_logging
functions. This helps avoid breaking too many conventions
regarding class naming schemes

WardBrian commented:
I think that the function-doubling-as-context-manager is fine and matches up well with some other patterns like how the `open` "function" works. pytorch's `no_grad` context manager also doubles as a decorator, which might be a step too far...

My other thoughts:
- It might be worth making [re-entrant](https://docs.python.org/3/library/contextlib.html#reentrant-cms) by having __entry__ set the `logger.disabled` flag again? I doubt this would come up in real use cases, but it doesn't hurt
- I would prefer a repr that returns `''`
- These two classes are almost identical, so I think it might be worth having one class that takes in a flag true/false, and then two functions that only create-and-return that class with a different argument.

bob-carpenter commented:
Thanks, @amas0.  I'm looking forward to the feature!

amas0 committed:
Combine EnableLogging/DisableLogging classes

amas0 committed:
Add empty string repr to ToggleLogging

amas0 committed:
Allow ToggleLogging to be reentrant

amas0 committed:
Add enable/disable logging unit tests

amas0 commented:
Okay, I believe I have addressed the points you raised @WardBrian. I added in some unit tests for good measure.

The re-entrant bit seems to work, as I've confirmed this monstrosity works fine:

```python
import cmdstanpy

dis = cmdstanpy.disable_logging()

with cmdstanpy.enable_logging() as en:
    with cmdstanpy.disable_logging():
        with en:
            with dis:
                model = cmdstanpy.CmdStanModel(stan_file="./bernoulli.stan")
                model.sample(show_progress=False, show_console=False, chains=1)
```

For a future 2.0, I think turning logging off by default with a context manager helper to turn it on would be a nice interface. Fiddling about with global logging state stuff is never particularly nice.

WardBrian commented on docsrc/users-guide/outputs.rst:
```diff
@@ -60,24 +60,47 @@ You may notice CmdStanPy can produce a lot of output when it is running:
+
+.. ipython:: python
+
+    cmdstanpy.enable_logging();
```
```suggestion
    cmdstanpy.enable_logging()
```

WardBrian commented on docsrc/users-guide/outputs.rst:
```diff
@@ -60,24 +60,47 @@ You may notice CmdStanPy can produce a lot of output when it is running:
-    cmdstanpy_logger = logging.getLogger("cmdstanpy")
-    cmdstanpy_logger.disabled = True
+    import cmdstanpy
+    cmdstanpy.disable_logging();
```
```suggestion
    cmdstanpy.disable_logging()
```

WardBrian commented on docsrc/users-guide/outputs.rst:
Would you mind also adding new sections and `.. autofunction::` declarations to `docsrc/api.rst`?

WardBrian commented on cmdstanpy/utils/logging.py:
```diff
@@ -1,25 +1,62 @@
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        traceback: Optional[types.TracebackType],
```
[optional] I realize these are just here to make typing happy, since we don't need them can we get away with `*args, **kwargs`? That way we don't need to import quite as much just for typing purposes

WardBrian reviewed (commented):
Thanks @amas0!

I have a few comments, mostly on the documentation, but otherwise this all looks good to me

amas0 committed:
Remove semicolons in output docs

Also fixup formatting

amas0 committed:
Add enable/disable_logging to api.rst

amas0 commented:
New commits should address all those points

WardBrian reviewed (approved):
Thanks @amas0!
